core/client/src/client.rs

```diff
@@ -1477,21 +1355,202 @@ where
 	Block: BlockT<Hash=H256>,
 {
+	/// Instantiate a new LongestChain for Backend B
+	pub fn new(backend: Arc<B>, import_lock: Arc<Mutex<()>>) -> Self {
+		LongestChain {
+			backend,
+			import_lock,
+			_phantom: Default::default()
+		}
+	}
+
 	fn best_block_header(&self) -> error::Result<<Block as BlockT>::Header> {
-		Client::best_block_header(self)
+		let info : ChainInfo<Block> = match self.backend.blockchain().info() {
+			Ok(i) => i,
+			Err(e) => return Err(error::Error::from_blockchain(Box::new(e)))
+		};
+		Ok(self.backend.blockchain().header(BlockId::Hash(info.best_hash))?
+			.expect("Best block header must always exist"))
+	}
+
+	/// Get the most recent block hash of the best (longest) chains
+	/// that contain block with the given `target_hash`.
+	///
+	/// The search space is always limited to blocks which are in the finalized
+	/// chain or descendents of it.
+	///
+	/// If `maybe_max_block_number` is `Some(max_block_number)`
+	/// the search is limited to block `numbers <= max_block_number`.
+	/// in other words as if there were no blocks greater `max_block_number`.
+	/// Returns `Ok(None)` if `target_hash` is not found in search space.
+	/// TODO: document time complexity of this, see [#1444](https://github.com/paritytech/substrate/issues/1444)
+	fn best_containing(
+		&self,
+		target_hash: Block::Hash,
+		maybe_max_number: Option<NumberFor<Block>>
+	) -> error::Result<Option<Block::Hash>> {
+		let target_header = {
+			match self.backend.blockchain().header(BlockId::Hash(target_hash))? {
+				Some(x) => x,
+				// target not in blockchain
+				None => { return Ok(None); },
+			}
+		};
+
+		if let Some(max_number) = maybe_max_number {
+			// target outside search range
+			if target_header.number() > &max_number {
+				return Ok(None);
+			}
+		}
+
+		let (leaves, best_already_checked) = {
+			// ensure no blocks are imported during this code block.
+			// an import could trigger a reorg which could change the canonical chain.
+			// we depend on the canonical chain staying the same during this code block.
+			let _import_lock = self.import_lock.lock();
+
+			let info = self.backend.blockchain().info()?;
+
+			let canon_hash = self.backend.blockchain().hash(*target_header.number())?
+				.ok_or_else(|| error::Error::from(format!("failed to get hash for block number {}", target_header.number())))?;
+
+			if canon_hash == target_hash {
+				// if no block at the given max depth exists fallback to the best block
+				if let Some(max_number) = maybe_max_number {
+					if let Some(header) = self.backend.blockchain().hash(max_number)? {
+						return Ok(Some(header));
+					}
+				}
+
+				return Ok(Some(info.best_hash));
+			} else if info.finalized_number >= *target_header.number() {
+				// header is on a dead fork.
+				return Ok(None);
+			}
+
+			(self.backend.blockchain().leaves()?, info.best_hash)
+		};
+
+		// for each chain. longest chain first. shortest last
+		for leaf_hash in leaves {
```

Comment:
This looks like it could be quadratic.

Reply:
Do you have a better algorithm?

Reply:
Not really, but we should make sure that this cannot be used in a denial-of-service attack.

Reply:
This PR focuses on splitting out the existing algorithm. I am fine with fixing it, but that is out of scope of this PR.